test/e2e/README.md

```diff
@@ -0,0 +1,112 @@
+# Crossplane End-to-end Tests
+
+This directory contains Crossplane's E2E tests. These tests exercise Crossplane
+features by:
+
+1. Creating a `kind` cluster.
+1. Deploying Crossplane, using its Helm chart.
+1. Applying a series of YAML manifests.
+1. Testing whether the resulting resources work as expected.
+
+These tests focus on core Crossplane features - i.e. the package manager,
+Composition, etc. They should not take dependencies on 'real' Crossplane
+providers or external systems. Instead they use 'fake' providers like
+`provider-nop` and `provider-dummy`.
+
+All Crossplane features must be exercised by these tests, as well as unit tests. 
+
+## Running Tests
+
+Run `make e2e` to run E2E tests.
+
+This compiles Crossplane and an E2E test binary. It then runs the test binary.
+Use the `E2E_TEST_FLAGS` to pass flags to the test binary. For example:
+
+```shell
+# Most tests use t.Log to explain what they're doing. Use the -test.v flag
+# (equivalent to go test -v) to see detailed test progress and logs.
+E2E_TEST_FLAGS="-test.v" make e2e
+
+# Some functions that setup the test environment (e.g. kind) use the klog logger
+# The -v flag controls the verbosity of klog. Use -v=4 for debug logging.
+E2E_TEST_FLAGS="-test.v -v=4" make e2e
+
+# To run only a specific test, match it by regular expression
+E2E_TEST_FLAGS="-test.run ^TestConfiguration" make e2e
+
+# To test features with certain labels, use the labels flag
+E2E_TEST_FLAGS="-labels area=apiextensions" make e2e
+```
+
+## Test Parallelism
+
+`make e2e` runs all defined E2E tests serially. Tests do not run in parallel.
+This is because all tests run against the same API server and Crossplane has a
+lot of cluster-scoped state - XRDs, Providers, Compositions, etc. It's easier
+and less error-prone to write tests when you don't have to worry about one test
+potentially conflicting with another - for example by installing the same
+provider another test would install.
+
+In order to achieve some parallelism at the CI level all tests are labelled with
+an area (e.g. `pkg`, `install`, `apiextensions`, etc). The [CI GitHub workflow]
+uses a matrix strategy to invoke each area as its own job, running in parallel.
+
+## Adding a Test
+
+Each feature under test consists of:
+
+1. A directory of manifests - i.e. test fixtures.
+1. A `features.Table` of assessment steps.
+
+Features are grouped into broad feature areas - e.g. `TestComposition` in
+`composition_test.go`. Features pertaining to Composition should be added to
+`TestComposition`.
+
+Some tests may involve updating an existing test - for example you might test a
+new kind of transform by updating the `composition/patch-and-transform`
+manifests and adding a new assessment or two to the `pandt` `features.Table`.
```

Comment:
Given we would require adding or updating e2e tests for every PR, I am curious how much the `features.Table` will grow, especially for features like Patch and Transform, where a lot of different cases need to be tested.

Reply:
I agree. For P&T specifically I was thinking we could perhaps test many patches and transforms as one 'feature' rather than breaking them all up. I.e. Feature == P&T rather than feature == toUpper.

I've added a note that this pattern isn't set in stone though - we're still learning. I do want us to discuss and evolve consciously though.